File: tests/XMP/global-view/template_fix/C-Round/gblock.c
#include <xmp.h>
#include <stdio.h>
#include <stdlib.h>     
static const int N=1000;
#pragma xmp nodes p(*)
#pragma xmp template t(:)
#pragma xmp distribute t(gblock(*)) onto p
int i,s=0,procs,remain;
int *a,*m,result=0;
#pragma xmp align a[i] with t(i)

int main(void)
{
  procs = xmp_num_nodes();
  m = (int *)malloc(sizeof(int) * procs);
  remain = N;

  for(i=0;i<procs-1;i++){
    m[i] = remain/2;
    remain -= m[i];
  }
  m[procs-1] = remain;

#pragma xmp template_fix(gblock(m)) t(0:N-1)
  a = (int *)xmp_malloc(xmp_desc_of(a), N);

#pragma xmp loop on t(i)
   for(i=0;i<N;i++)
     a[i] = i;

#pragma xmp loop on t(i) reduction(+:s)
   for(i=0;i<N;i++)
     s += a[i];
   
   if(s != 499500)
     result = -1;

#pragma xmp reduction(+:result)
#pragma xmp task on p(1)
   {
     if(result == 0){
       printf("PASS\n");
     }
     else{
       fprintf(stderr, "ERROR\n");
       exit(1);
     }
   }
   
   free (a);
   free (m);
   return 0;
}

File: tests/XMP/global-view/template_fix/F/gblock.f90
program main
  include 'xmp_lib.h'
  integer,parameter:: N=1000
!$xmp nodes p(*)
!$xmp template t(:)
!$xmp distribute t(gblock(*)) onto p
  integer s, procs, remain, result
  integer,allocatable:: a(:)
  integer,allocatable:: m(:)
!$xmp align a(i) with t(i)
  
  procs = xmp_num_nodes()
  allocate(m(procs))
      
  remain = N
  do i=1, procs-1
     m(i) = remain/2
     remain = remain-m(i)
  enddo
  m(procs) = remain

!$xmp template_fix (gblock(m)) t(N)
  allocate(a(N))

!$xmp loop (i) on t(i)
  do i=1, N
     a(i) = i
  enddo
  
  s = 0
!$xmp loop (i) on t(i) reduction(+: s)
  do i=1, N
     s = s+a(i)
  enddo
  
  result = 0
  if(s .ne. 500500) then
     result = -1
  endif

!$xmp reduction(+:result)
!$xmp task on p(1)
  if( result .eq. 0 ) then
     write(*,*) "PASS"
  else
     write(*,*) "ERROR"
     call exit(1)
  endif
!$xmp end task

  deallocate(a, m)

end program main
